### ISSUE TITLE
Value Checker range refinement not working in a loop

### ISSUE THREAD

**Issue by panacekcz** [typetools/checker-framework#1423]
In this code, I would expect the Value Checker to refine the range of  variable `i` and report no warnings. However, it does not refine the range in a loop containing an increment of another variable:
```java
import org.checkerframework.common.value.qual.IntRange;
public class RangeRefinement {
	void loop(int i) {
		int a = 0;
		while (i >= 2) {
			@IntRange(from = 2) int i2 = i;
			++a;
		}
	}
}
```
Value checker error output:
```
RangeRefinement.java:6: error: [assignment.type.incompatible] incompatible types in assignment.
			@IntRange(from = 2) int i2 = i;
			                             ^
  found   : @IntRange(from=-2147483648, to=2147483647) int
  required: @IntRange(from=2, to=2147483647) int
1 error
```

Each of these changes make the error go away:
- change `while` to `if`
- change `>=` to `==`
- remove `++a`
- change `++a` to `a*=2`
- change `++a` to `++i`
- change `++a` to `a+=i`

On the other hand, changing `++a` to for example `a-=100` does not.

This issue applied to string length refinement causes ~~panacekcz#6~~ #1432 not to work in [this method](https://github.com/mernst/plume-lib-check-index-new/blob/3f82f73faaf4c01bee92944a1ea16d8e854bf2fb/java/src/plume/BCELUtil.java#L664) in plume-lib.

**Comment by kelloggm** [comment]
Could this be related to #867?

**Comment by smillst** [comment]
Commit ec42d62bc19e09b2ba283a5e2b0c185aff157b9c improves refinement in a loop a bit.  Your example now outputs:
```
RangeRefinement.java:6: error: [assignment.type.incompatible] incompatible types in assignment.
			@IntRange(from = 2) int i2 = i;
			                             ^
  found   : @IntRange(from=-128, to=2147483647) int
  required: @IntRange(from=2, to=2147483647) int
```

The dataflow analysis iterates over the code in a loop until a fixed point is reached -- the types stop changing.  Because the Value Checker is actually computing values in a loop, this could take a long time or possible not terminate.  To prevent this, the Value Checker 
uses `ValueAnnotatedTypeFactory.ValueQualifierHierarchy#widenedUpperBound` to estimate the value rather than compute it.  I tried to various implementations of this method to improve the estimate, but other implementations cause infinite loops in some cases.

**Comment by panacekcz** [comment]
@smillst, do you know what causes the infinite loops? Do you think would be possible to change something to enable having more precise widening in the value annotation hierarchy without causing infinite loops?

**Comment by smillst** [comment]
I don't think there is an easy or straight forward way to improve this. One problem is that the dataflow framework doesn't identify which types in the store are changing, so all types in the store have to be widened.  I didn't see way to implement this without a major rewrite.

**Comment by panacekcz** [comment]
I tried a simple improvement to widening in the value checker, which makes it work for this example: https://github.com/typetools/checker-framework/compare/master...panacekcz:bounded-strings-widening

I think that when I first tried it, I encountered an infinite loop in the index checker, but I can't reproduce it anymore, so I am not really sure what happened there. Now, both local and travis build of this version succeeds for me.
